eclipse-theia/theia
# core,mini-browser: server enhancements

# mini-browser: serve on separate origin

The mini-browser currently hosts its services on the same origin than
Theia's main origin. This commit makes the `mini-browser` serve on its
own origin: `{{uuid}}.mini-browser.{{hostname}}` by default. Can be
configured with a `THEIA_MINI_BROWSER_HOST_PATTERN` environment
variable.

# core: validate ws upgrade origin

Hosting the `mini-browser` on its own origin prevents cross-origin
requests from happening easily, but WebSockets don't benefit from the
same protection. We need to allow/disallow upgrade requests in the
backend application ourselves.

This means that in order to know who to refuse, we need to know where we
are hosted. This is done by specifying the `THEIA_HOSTS` environment
variable. If left empty, no check will be done. But if set, nothing
besides what is written in this var will be allowed. See
`BackendApplicationHosts` to get the hosts extracted from this var. Note
that the latter is not set when running Electron, since there's no need
to deal with arbitrary origins, everything should be local.

No check is done on the HTTP(s) endpoints because we'll rely on the
browser's SOP and CORS policies to take effect.

A new contribution point is added: `WsRequestValidatorContribution`.
An extender can implement this contribution to allow or not WebSocket
connections. Internally used to filter origins and Electron's token as
well as checking the origin of requests to prevent some type of XSS.

Signed-off-by: Paul Maréchal <paul.marechal@ericsson.com>

Fixes: https://bugs.eclipse.org/bugs/show_bug.cgi?id=568018

#### How to test

- The `mini-browser` should work like before:
  - Preview a `.html` file, open the dev-tools and make sure that the file is served from `mini-browser.localhost:3000`.
  - Copy the url of the `iframe` and open it in a new tab.
  - Open the dev-tools and try opening a WebSocket connection to the main Theia application on `ws://localhost:3000/services`.
  - It should fail.
- The backend should be able to refuse connections from wrong hosts:
  - Shutdown the backend, set `THEIA_HOSTS=patate.local:3000`.
  - Edit your `hosts` file to add an entry like `127.0.0.1 patate.local`.
  - Run the backend and try accessing the app on http://localhost:3000.
  - It should not work due to the WebSocket connection being refused.
  - Access the app on http://patate.local:3000.
  - It should work.
- Electron should work like before:
  - Should not be affected by `THEIA_HOSTS`.

#### Review checklist


#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)
